### ISSUE TITLE
Dashboard: author does not see drafts

### ISSUE THREAD

**Issue by BrittanyIRL** [GoogleForCreators/web-stories-wp#6613]
## Bug Description

If I log in to an established wordPress account as a user with no stories of my own but another user has created stories, those stories aren't showing up in "my stories" (because they aren't mine) however the header counts for existing stories are showing with the number of stories that exist for that other user.  

![Screen Shot 2021-03-04 at 8.26.33 AM.png](https://images.zenhubusercontent.com/5e9dcd9a9ff5dfdaffcdf132/d8d69ebf-d3cf-455f-b3d9-10679dffe878)

## Expected Behaviour

If I have no stories that are getting returned to me I would expect the story counts to all be 0 regardless of what other account users have created.

## Steps to Reproduce

Log into a wordPress account that has stories created in it but as either a new users or a user with 0 stories authored by them, go to the dashboard. See header counts. 

## Additional Context

returned headers from `web-stories/v1/web-story/` are: `"headers":{"X-WP-Total":18,"X-WP-TotalPages":1,"X-WP-TotalByStatus":"{\"all\":20,\"publish\":0,\"future\":0,\"draft\":18,\"private\":2}"}}`

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

That header counts are 0 when there are no stories to show.

**Comment by swissspidy** [comment]
> those stories aren't showing up in "my stories"

Since when are we filtering "My Stories" so that it only actually displays the current user's stories? 🤔 

Aren't we displaying _all stories_, we just call it "My Stories" in the UI?

What is the user role of your user?

**Comment by BrittanyIRL** [comment]
> What is the user role of your user? 

all of the stories in my local environment are under the 'admin'. this happens when i log into the existing local environment with my author user role.

**Comment by BrittanyIRL** [comment]
> Since when are we filtering "My Stories" so that it only actually displays the current user's stories

Good question, when logged in with and admin role I can see stories authored by anyone. So it's not "my stories". Is it the author role? I'm not sure where specifically the disconnect is, but the page presented should be consistent.

**Comment by swissspidy** [comment]
Yeah so users with role **Author** cannot edit other people's posts.

Users with role **Administrator** and **Editor** can do that.

Let me rephrase the ticket to make that more clear.

**Comment by swissspidy** [comment]
Actually, I cannot reproduce this.

Logging in as an author I see all stories in the dashboard still. Same as on the "All Stories" view.

**Comment by BrittanyIRL** [comment]
@swissspidy Does your author have stories in the above situation?

**Comment by swissspidy** [comment]
> @swissspidy Does your author have stories in the above situation?

No, that account doesn't have any stories. They were all created by the admin.

**Comment by BrittanyIRL** [comment]
Huh, that's odd. 

Something weird is definitely going on, had @embarks and @samwhale also look locally in their environments. Em has the same set up as me, Samuel is seeing what you are seeing. 

My results in stories as an author role with 1 story authored by author:
`Request URL: http://localhost:8899/wp-json/web-stories/v1/web-story/?_embed=author&_web_stories_envelope=true&context=edit&order=asc&orderby=title&page=1&per_page=24&status=publish%2Cdraft%2Cfuture%2Cprivate&_locale=user` 
![Screen Shot 2021-03-04 at 10 25 04 AM](https://user-images.githubusercontent.com/10720454/110004121-51125900-7cd4-11eb-934b-d69a9ea5e518.png)

This should show 20 more items in `body` than it does to have the proper story count that matches 'all stories' view by wordpress. The header area with the counts of how many drafts, published and total stories are not matching what is getting returned by the API. 


https://user-images.githubusercontent.com/10720454/110005964-4c4ea480-7cd6-11eb-8fce-7450a0701dcd.mp4

**Comment by swissspidy** [comment]
@spacedmonkey perhaps you are able to reproduce this?

**Comment by swissspidy** [comment]
Looks like the issue here is that @BrittanyIRL does only have drafts. And author users don't have permission to read or edit someone else's drafts in WordPress.

**Scenario**

7 posts created by admin (6 published, 1 draft)

**All Stories View**

Author user sees all 7 stories. The 6 published ones can only be viewed on the frontend (not edited), the draft one can neither be edited nor viewed

**My Stories View**

Author user sees only the 6 published stories, not the draft one (yet the count says 7).

Each story incorrectly has a "Open in editor" link, which will lead you to an error screen

<img width="829" alt="Screenshot 2021-03-08 at 10 10 05" src="https://user-images.githubusercontent.com/841956/110299819-7534a980-7ff6-11eb-814c-94a482d4005f.png">

So I see two issues here:

1. The stories have an "Open in editor" link, even though they cannot actually be edited  
   Same goes for "Rename", "Duplicate", "Delete". Author users cannot do that for someone else's story
2. The draft stories are not visible to the author  
   Question: should the author be allowed to see the admin draft story's preview? Probably not.

To me that seems like a pea pod issue. Although might need some WP pod help for ensuring the permissions can be checked appropriately.

**Comment by spacedmonkey** [comment]
Sorry a came a little late to this one. The dashboard does not seem to respect `edit_others_posts` or `delete_others_posts` use capabilities. See [get_post_type_capabilities](https://developer.wordpress.org/reference/functions/get_post_type_capabilities/#description) details for a list of capabilities. 

Adding these capabilities to the config dashboard here 

https://github.com/google/web-stories-wp/blob/3532692b15da22809e6e597367f252a8f18724e0/includes/Dashboard.php#L341-L347

https://github.com/google/web-stories-wp/blob/3532692b15da22809e6e597367f252a8f18724e0/includes/Dashboard.php#L372-L376

Once we have these in the config, we should check if a user can edit / delete others stories before displaying this options to the user. 

It is worth noting the rest api response different from what is expected in the WP screen. The REST API only returns stories you can edit. The WP screen, puts placeholder lines,  with no title so the number of results match lines in table. Compare these screenshots. 

<img width="1884" alt="Screenshot 2021-03-08 at 12 15 22" src="https://user-images.githubusercontent.com/237508/110320453-37408100-8008-11eb-8149-0fac7454a5a0.png">
<img width="1884" alt="Screenshot 2021-03-08 at 12 16 02" src="https://user-images.githubusercontent.com/237508/110320465-3dcef880-8008-11eb-89df-c4bdf4d29506.png">

So what is the expected behavior? 
1. The number of drafts returned in API header, reflect the user that user can see? 
2. Or should we put a number of placeholder stories to fill out this screen so the numbers match?

IMO, I think option 1. 

Sounds like we ( infra ) need to add some extra data to config for capabilities and add more logic to rest api to check for `edit_others_posts` and change the counts.

**Comment by swissspidy** [comment]
> It is worth noting the rest api response different from what is expected in the WP screen. The REST API only returns stories you can edit. The WP screen, puts placeholder lines, with no title so the number of results match lines in table. Compare these screenshots.

Didn't realize that's how the REST API works.

In that case, is it possible to easily exclude these from the count?

**Comment by BrittanyIRL** [comment]
Thanks for making sense of what is going on, @spacedmonkey ! Seems like this is two fold. `) if we can easily exclude to arrive at the suggested expected behavior:
> The number of drafts returned in API header, reflect the user that user can see?

Two, front end needs to check for permissions before setting up actions available on stories.  

If this is accurate, I'll make a new bug ticket for pea pod for that and ya'll can have this one for the count update.

**Comment by spacedmonkey** [comment]
I just want to confirm my thinking here on the course forward. 

- Add `user_id`, `edit_others_posts` and `delete_others_posts` to the config object used by dashboard. ( Infra )
- Add a check to compare the current user id with author of the story and the `edit_others_posts` capability. If the user do not have `edit_others_posts` capability and was not the author of the story, then all options to edit the story should be remove. This links to edit or preview and the ability to change the title of the story should also be disabled. This would not currently work anyway, as the REST API would reject it. 
- Similar to the above, checks should be in place for `delete_others_posts` and `delete_posts` and options to delete stories should be hidden. 
- Tweak the headers returned in REST API, if author ( or any other users without `edit_others_posts` or `delete_others_posts` ) the counts should be changed to reflect what a user can edit. 

I would break this into 3 tickets. 
- Changes to the REST API 
- Add capabilities to dashboard object. 
- Change dashboard js to respect `edit_others_posts` and `delete_others_posts` capabilities.

**Comment by BrittanyIRL** [comment]
@spacedmonkey I think the above that you suggest would be a great course forward. I see this as 2 infra tickets (first two in your list) and 1 pea pod ticket (last item in list).

**Comment by swissspidy** [comment]
@spacedmonkey I'm not sure we should be do it this way (passing to config and comparing user IDs etc).

Can't we infer the capability from the REST API response (the `_links` part)? That'd be the proper way.

Take `hasAssignAuthorAction` as an example. There is a `wp:action-assign-author` entry under `_links` if the user has the capability. We should be checking for that instead of passing that manually via config.

Same goes for `edit_others_posts` checks.

> Tweak the headers returned in REST API, if author ( or any other users without edit_others_posts or delete_others_posts ) the counts should be changed to reflect what a user can edit.

As per my previous question, is this easily possible? How could this be done? This is still unclear to me.

---

Let's hold off on creating tickets until we've come to an agreement regarding implementation.

**Comment by spacedmonkey** [comment]
> Take hasAssignAuthorAction as an example. There is a wp:action-assign-author entry under _links if the user has the capability. We should be checking for that instead of passing that manually via config

Can instead, see this [snippet](https://github.com/WordPress/wordpress-develop/blob/6877f57f368df6850732a1e3c6144dc1973f3cc1/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2066-L2070). However, there is no simple way to check for  `delete_others_posts`. 

Considering that already have capability checks in config objection, I don't see the harm in adding more. 
https://github.com/google/web-stories-wp/blob/80d7f526103ca1396ce3eb4d40a2bc91c277c302/includes/Dashboard.php#L358-L362

> As per my previous question, is this easily possible? How could this be done? This is still unclear to me.

I haven't worked this out exactly either but it would look something like this. 

```
		// Add counts for other statuses.
		$statuses = [
			'publish' => 'publish',
			'future'  => 'future',
			'draft'   => 'draft',
			'private' => 'private',
		];

		$statuses_count = [];

		// Strip down query for speed.
		$query_args['fields']                 = 'ids';
		$query_args['posts_per_page']         = 1;
		$query_args['update_post_meta_cache'] = false;
		$query_args['update_post_term_cache'] = false;

                $post_type_object         = get_post_type_object( Story_Post_Type::POST_TYPE_SLUG );

		foreach ( $statuses as $key => $status ) {
			$posts_query               = new WP_Query();
			$query_args['post_status'] = $status;
                         if ( 'draft' === $status && current_user_can( $post_type_object->cap->edit_others_posts ) ) {
                               $query_args['post_author'] = get_current_user_id();
                         }
			$posts_query->query( $query_args );
			$statuses_count[ $key ] = absint( $posts_query->found_posts );
		}

                 $statuses_count[ 'all' ]  = array_sum ( array_values ( $statuses_count ) );
```

**Comment by swissspidy** [comment]
This somehow slipped through, apologies for the late response.

@spacedmonkey Could you perhaps give this a try this week to see whether the count works as expected with the method you've shared?

Still don't agree with the `Add a check to compare the current user id with author of the story` part as this shouldn't be necessary to do on the client side. We should use REST API `_links` instead in the future.

Note: we don't need to fix this ticket this week, but getting a better understanding of it and an implementation brief would be a great outcome.

**Comment by swissspidy** [comment]
@spacedmonkey friendly ping, curious to hear your thoughts on fixing the count

**Comment by swissspidy** [comment]
Is this still an issue after #8598 landed?

**Comment by BrittanyIRL** [comment]
> Is this still an issue after #8598 landed?

i'll check today

**Comment by BrittanyIRL** [comment]
Still an issue. Seems possible it's related to what @littlemilkstudio mentioned in #8532

**Comment by csossi** [comment]
Logged in as an Author and seeing the totals off for "All Stories" and "Drafts":
all stories
![image](https://user-images.githubusercontent.com/16945664/134034120-d19407f0-ff43-4777-b78c-2340a083062c.png)
drafts
![image](https://user-images.githubusercontent.com/16945664/134034218-dc6e4c25-f972-490e-94c1-ee8e64ae72df.png)

I'm also seeing the top nav change as I scroll down the page under "All Stories" when logged in as an admin. In the video, note the top nav selections at start and then selections available at the end ("Scheduled" and "Private" disappear). You can also see that spinners are appearing (according to QA Instructions, they shouldn't):
https://share.getcloudapp.com/6quYA58O
